File: api/src/main/java/fr/huiitre/tools/modules/core/notification/application/usecase/SendNotificationUseCase.java
package fr.huiitre.tools.modules.core.notification.application.usecase;

import org.springframework.context.ApplicationEventPublisher;
import org.springframework.stereotype.Service;

import fr.huiitre.tools.modules.core.notification.application.event.NotificationEvent;
import fr.huiitre.tools.modules.core.role.domain.RoleCode;
import fr.huiitre.tools.modules.core.security.application.usecase.SecuredUseCase;

@Service
public class SendNotificationUseCase implements SecuredUseCase {

    private final ApplicationEventPublisher eventPublisher;

    public SendNotificationUseCase(ApplicationEventPublisher eventPublisher) {
        this.eventPublisher = eventPublisher;
    }

    @Override
    public RoleCode requiredRole() {
        return RoleCode.TECH;
    }

    public void execute(SendNotificationCommand command) {
        NotificationEvent event = new NotificationEvent(
            command.title(),
            command.body(),
            command.type(),
            command.targetUserId(),
            command.targetRoleId(),
            command.targetModuleId(),
            command.metadata()
        );
        eventPublisher.publishEvent(event);
    }
}

File: api/Modules/Core/Notifications/Application/Usecases/SendNotificationUseCase.cs
using Tools.Api.Modules.Core.Notifications.Application.Services;
using Tools.Api.Modules.Core.Security.Application.Services;
using Tools.Api.Modules.Core.Security.Application.Usecases;
using Tools.Api.Modules.Core.Security.Domain;

namespace Tools.Api.Modules.Core.Notifications.Application.Usecases;

// Envoi manuel d'une notification, déclenché par un humain.
//
// TECH, comme dans l'API Java : c'est un outil d'exploitation, pas une action de produit. À ne
// pas confondre avec `PublishInternalNotificationUseCase`, qui sert les appels de service à
// service et n'a aucun utilisateur à autoriser.
public sealed class SendNotificationUseCase(
    UseCaseAuthorizer authorizer,
    NotificationService notificationService) : SecuredUseCase(authorizer)
{
    protected override RoleCode RequiredRole => RoleCode.Tech;

    // Retourne l'identifiant créé, nul si la cible n'a désigné aucun destinataire.
    public Task<long?> Execute(SendNotificationCommand command)
    {
        return notificationService.Send(command);
    }
}
